src/lib/cache.js

```diff
@@ -62,53 +62,46 @@ function createRedisClient() {
   }
   client.on("error", error)
   VerboseEvents.forEach(event => {
-    client.on(event, () => verbose(`Redis: ${event}`))
+    client.on(event, () => verbose(`[Cache] ${event}`))
   })
   return client
 }
 
 export const client = isTest ? createMockClient() : createRedisClient()
 
-// Used to timeout Redis calls
-// From: https://medium.com/@stockholmux/tick-tock-setting-a-timeout-on-a-redis-call-with-node-js-be63733df98a
-function timeoutRedisCommand(command, timeoutValue) {
-  return function wrappedCommand() {
-    const argsAsArray = Array.prototype.slice.call(arguments)
-    let callback = argsAsArray.pop()
-    const timeoutHandler = setTimeout(() => {
-      error(`Cache ${command} timeout`)
-      callback(new Error(`Cache ${command} timeout`))
-      callback = () => {}
-    }, timeoutValue)
-
-    argsAsArray.push((err, data) => {
-      clearTimeout(timeoutHandler)
-      callback(err, data)
-    })
-
-    client[command].apply(client, argsAsArray)
-  }
-}
-
-const getWithTimeout = timeoutRedisCommand("get", CACHE_RETRIEVAL_TIMEOUT_MS)
-
 export default {
   get: key => {
     return new Promise((resolve, reject) => {
-      if (isNull(client)) return reject(new Error("Cache client is `null`"))
+      if (isNull(client)) return reject(new Error("[Cache] `client` is `null`"))
+
+      let timeoutId = setTimeout(() => {
+        timeoutId = null
```

Comment:
Is there much difference just using a `const` and not bothering to unset `timeoutId`? From a quick experiment in my REPL, there doesn't seem to be any risk to calling `clearTimeout` if the timeout has already triggered. But I was curious whether there are hazards I'm missing.

Reply:
> From a quick experiment in my REPL, there doesn't seem to be any risk to calling `clearTimeout`

Ah nice, I had not tested that 👌 

However, in this case I am mostly using the `timeoutId` being `null` as a sentinel to know that the execution timed out and needing to continue past (line 95](https://github.com/artsy/metaphysics/pull/1060/files#diff-9ead3d77f16d7f3a8c380c297fa056e4R95).

Reply:
Ah gotcha, I see now. That's a bit non-obvious, but I can't really think of an easy refactor that would be more straightforward. If I think of something, I'll follow up.

Reply:
```javascript
function cancellableDelay(delay, value = {}) {
  let timeoutId
  const promise = new Promise((resolve, reject) => {
    timeoutId = setTimeout(() => resolve(value), delay)
  })
  const cancel = () => clearTimeout(timeoutId)
  return [promise, cancel]
}
```

Then I'd make a promise-ifieid `client.get`:

```javascript
function getKey(key) {
  return new Promise((resolve, reject) => {
    client.get(key, (err, data) => err ? reject({ getKeyError: error}) : resolve(data))
  })
}
```

Then everything could be composed together:

```javascript
export default {
  async get(key) {
    if (isNull(client)) throw new Error("[Cache] `client` is `null`")

    const [timeoutPromise, cancelTimeout] = cancellableDelay(
      CACHE_RETRIEVAL_TIMEOUT_MS,
      { timeout: `[Cache#get] Slow read of ${time}ms for key ${key}` }
    )

    const start = Date.now()

    const cachePromise = getKey(key)

    try {
      const data = await Promise.race([ cachePromise, timeoutPromise ])

      const time = Date.now() - start
      if (time > CACHE_QUERY_LOGGING_THRESHOLD_MS) {
        error(`[Cache#get] Slow read of ${time}ms for key ${key}`)
      }

      if (data) {
        return JSON.parse(data)
      } else {
        throw { cacheMiss: "[Cache#get] Cache miss" }
      }
    } catch (err) {
      const { timeout, cacheMiss, getKeyError } = err
      if (timeout) {
        error(timeout)
        throw new Error(timeout)
      } else if (cacheMiss) {
        throw new Error(cacheMiss)
      } else if (getKeyError) {
        error(getKeyError)
        throw new Error(getKeyError)
      } else {
        // Presumably, any other cause is already an Error, so re-throw
        error(err)
        throw err
      }
    } finally {
      cancelTimeout()
    }
  }
}
```

I haven't tested that, but I think the only functional difference is that it won't log slow reads on gets that error out. But if that's important, it could be added back to that part of the error handler.

If that actually works, with individual async actions tightly encapsulated in promises, I suspect it would end up easier to add more complex behavior, like [speculative retry](https://bjankie1.github.io/blog/), should we decide to.

Reply:
@acjay The redis client already comes wit promises, but I did not use that last time there was an incident and we switched back to redis as I wanted to change as little as possible other code. Now seems a good time to iterate on that, though, so please go for it because I agree that it looks like we could move forward more easily with yours 👍